File: Medium/396_Rotate_Function/Solution.py
from typing import List


# class Solution:
#     def maxRotateFunction(self, nums: List[int]) -> int:
#         n = len(nums)
#         nums = nums + nums
#         maxNum = float('-inf')

#         for i in range(n):
#             j = i
#             m = 0
#             total = 0

#             while j < i + n:
#                 total += nums[j] * m
#                 m += 1
#                 j += 1
            
#             maxNum = max(maxNum, total)
        
#         return maxNum


class Solution:
    def maxRotateFunction(self, nums: List[int]) -> int:
        n = len(nums)
        totalSum = sum(nums)

        F = sum(i * nums[i] for i in range(n))
        maxVal = F

        for k in range(1, n):
            F = F + totalSum - n * nums[n - k]
            maxVal = max(maxVal, F)

        return maxVal


if __name__ == "__main__":
    solution = Solution()

    # test cases 1
    nums1 = [4, 3, 2, 6]
    print(solution.maxRotateFunction(nums1))

    # test cases 2
    nums2 = [100]
    print(solution.maxRotateFunction(nums2))

    # test cases 1
    # nums3 = [-4, -3, -2, -6]
    # print(solution.maxRotateFunction(nums3))

File: Medium/396_Rotate_Function/Solution.cpp
#include "../../devlibs/cpp/cpphelper.h"

using namespace std;


// NOTE:
// For LeetCode submission, copy only the `class Solution` part.


class Solution {
public:
    int maxRotateFunction(vector<int>& nums) {
        int n = nums.size();
        long long totalSum = 0;
        long long tmp = 0;
        long long maxVal = 0;

        for (int i = 0; i < n; i++) {
            totalSum = totalSum + nums[i];
        }

        for (int i = 0; i < n; i++) {
            tmp = tmp + (i * nums[i]);
        }

        maxVal = tmp;

        for (int i = 1; i < n; i++) {
            tmp = tmp + totalSum - n * nums[n - i];
            maxVal = max(maxVal, tmp);
        }

        return maxVal;
    }
};


int main(int argc, char* argv[]) {
    ios::sync_with_stdio(false);
    cin.tie(NULL);

    Solution solution;
    
    // test cases 1
    vector<int> nums1 = {4, 3, 2, 6};
    cout << solution.maxRotateFunction(nums1) << endl;

    // test cases 2
    vector<int> nums2 = {100};
    cout << solution.maxRotateFunction(nums2) << endl;

    return 0;
}